src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -880,8 +940,11 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                             {
                                 throw new IOException(SR.net_ssl_io_renego);
                             }
-
                             await ReplyOnReAuthenticationAsync(adapter, extraBuffer).ConfigureAwait(false);
+                            if (renegotiation)
```

Comment:
It still feels like there is a case missing here:

If we receive a frame that contains actual data (not a response to renegotiation), then we should throw, right? As we do above if _decryptedBytesCount != 0?

Reply:
that is correct.  In this version of `ReadAsyncInternal`, `DecryptData`  is called and if there is no error it would loop back again and hit the check.

Reply:
Okay -- can we add a comment somewhere explaining this? It was not obvious to me.

Reply:
yes, I can. This will also change once #50815 goes in.